crates/red_knot_python_semantic/src/types.rs

```diff
@@ -3035,47 +3116,126 @@ pub enum TypeVarBoundOrConstraints<'db> {
-    PossiblyUnboundDunderIter {
+enum IterateError<'db> {
+    /// The type isn't iterable because it doesn't implement the `__iter__` or `__get__item` "protocol".
+    NotIterable { not_iterable_ty: Type<'db> },
+
+    /// The type is iterable but the methods aren't always bound.
+    PossiblyUnbound {
         iterable_ty: Type<'db>,
         element_ty: Type<'db>,
     },
 }
 
-impl<'db> IterationOutcome<'db> {
-    fn unwrap_with_diagnostic(
-        self,
-        context: &InferContext<'db>,
-        iterable_node: ast::AnyNodeRef,
-    ) -> Type<'db> {
+impl<'db> IterateError<'db> {
+    /// Reports the diagnostic for this error.
+    fn report_diagnostic(&self, context: &InferContext<'db>, iterable_node: ast::AnyNodeRef) {
         match self {
-            Self::Iterable { element_ty } => element_ty,
             Self::NotIterable { not_iterable_ty } => {
-                report_not_iterable(context, iterable_node, not_iterable_ty);
-                Type::unknown()
+                report_not_iterable(context, iterable_node, *not_iterable_ty);
             }
-            Self::PossiblyUnboundDunderIter {
+            Self::PossiblyUnbound {
                 iterable_ty,
-                element_ty,
+                element_ty: _,
             } => {
-                report_not_iterable_possibly_unbound(context, iterable_node, iterable_ty);
-                element_ty
+                report_not_iterable_possibly_unbound(context, iterable_node, *iterable_ty);
             }
         }
     }
 
-    fn unwrap_without_diagnostic(self) -> Type<'db> {
+    /// Returns the element type if it is known, or `None` if the type is never iterable.
+    fn element_type(&self) -> Option<Type<'db>> {
+        match self {
+            IterateError::NotIterable { .. } => None,
+            IterateError::PossiblyUnbound { element_ty, .. } => Some(*element_ty),
+        }
+    }
+
+    /// Returns the element type if it is known, or `Type::unknown()` if it is not.
+    fn fallback_element_type(&self) -> Type<'db> {
+        self.element_type().unwrap_or(Type::unknown())
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub(super) enum BoolError<'db> {
+    /// The type has a `__bool__` attribute but it can't be called.
+    NotCallable { not_boolable_type: Type<'db> },
+
+    /// The type has a callable `__bool__` attribute, but it doesn't have the
+    /// expected signature (parameters or return type).
+    BindingError {
+        not_boolable_type: Type<'db>,
+        return_type: Type<'db>,
+    },
+
+    /// Any other reason why the type can't be converted to a bool.
+    /// E.g. because the type is a union and not all variants support `__bool__` or
+    /// because `__bool__` points to a type that has a possibly unbound `__call__` method.
+    Other { not_boolable_type: Type<'db> },
+}
+
+impl BoolError<'_> {
+    #[allow(clippy::unused_self)]
+    pub(super) fn fallback_truthiness(&self) -> Truthiness {
+        Truthiness::Ambiguous
+    }
```

Comment:
hrm, I think I'd prefer to add a `Truthiness::from_bool_error()` function and use that

Reply:
Yeah, I see the consistency argument... I still feel like I'd prefer something inconsistent here than having a method that "pretends to be an instance method" but doesn't actually use `self`. That said, I don't feel very strongly.

Could we at least add a comment about why we're deliberately leaving `self` unused here, if we leave it as-is, though?